<?xml version="1.0" encoding="utf-8"?>
<!--
Launch file for Selective ICP with rviz
-->

<launch>

  <!-- Proper Output Info -->
  <env name="ROSCONSOLE_FORMAT" value="[${severity}] (${node}) [${time}]: ${message}\n" />

  <param name="/use_sim_time" value="true" />
    <arg name="cad_file" default="$(find cpt_selective_icp)/resources/cla.json" />
    <arg name="rosbag" />
    <arg name="logprefix" />
    <arg name="segmentation_model" />
    <arg name="input_height" />
    <arg name="input_width" />
    <arg name="threshold" default="150" />

 <!--node pkg="rosbag"
        type="record"
        name="output_recorder"
        args="-o /tmp/bfseg_output_$(arg logprefix)  /mesh_publisher/mesh_out /leica/position /rslidar_points /points_filtered /pickelhaube_resized/cam0/image /pickelhaube_resized/cam0/segmented /pickelhaube_resized/cam1/image /pickelhaube_resized/cam1/segmented /pickelhaube_resized/cam2/image /pickelhaube_resized/cam2/segmented /tf /tf_static"
        /-->

    <node pkg="background_foreground_segmentation"
          type="pose_logger.py"
          name="poselogger">
        <param name="filename" value="/home/matthias/data/rss_2021_data/$(arg logprefix)_icp.csv" />
    </node>
    <node pkg="background_foreground_segmentation"
          type="point_logger.py"
          name="leicalogger">
        <param name="filename" value="/home/matthias/data/rss_2021_data/$(arg logprefix)_leica.csv" />
        <param name="overwrite" value="false" />
    </node>

    <!-- Start Dataset Creator node for each camera -->
    <node pkg="background_foreground_segmentation" type="background_foreground_segmentation_node2"
          name="dataset_creator_cam0" output="screen">
        <rosparam command="load" file="$(find background_foreground_segmentation)/config/online_dataset_creator.yaml"/>
        <param name="outputFolder" value="/home/matthias/data/rss_2021_data/pseudolabels/"/>
        <param name="cameraImageTopic" value="/pickelhaube_resized/cam0/image"/>
        <param name="cameraInfoTopic" value="/pickelhaube_resized/cam0/camera_info"/>
        <param name="cameraFrame" value="/cam0_rectified_frame"/>
        <param name="outputTopic" value="/meshdist/cam0/original"/>
        <param name="labelsTopic" value="/meshdist/cam0/projected_pc"/>
        <param name="distanceTopic" value="/meshdist/cam0/projected_pc_distance"/>
    </node>

    <node pkg="background_foreground_segmentation" type="background_foreground_segmentation_node2"
          name="dataset_creator_cam1" output="screen">
        <rosparam command="load" file="$(find background_foreground_segmentation)/config/online_dataset_creator.yaml"/>
        <param name="outputFolder" value="/home/matthias/data/rss_2021_data/pseudolabels/"/>
        <param name="cameraImageTopic" value="/pickelhaube_resized/cam1/image"/>
        <param name="cameraInfoTopic" value="/pickelhaube_resized/cam1/camera_info"/>
        <param name="cameraFrame" value="/cam1_rectified_frame"/>
        <param name="outputTopic" value="/meshdist/cam1/original"/>
        <param name="labelsTopic" value="/meshdist/cam1/projected_pc"/>
        <param name="distanceTopic" value="/meshdist/cam1/projected_pc_distance"/>
    </node>

    <node pkg="background_foreground_segmentation" type="background_foreground_segmentation_node2"
          name="dataset_creator_cam2" output="screen">
        <rosparam command="load" file="$(find background_foreground_segmentation)/config/online_dataset_creator.yaml"/>
        <param name="outputFolder" value="/home/matthias/data/rss_2021_data/pseudolabels/"/>
        <param name="cameraImageTopic" value="/pickelhaube_resized/cam2/image"/>
        <param name="cameraInfoTopic" value="/pickelhaube_resized/cam2/camera_info"/>
        <param name="cameraFrame" value="/cam2_rectified_frame"/>
        <param name="outputTopic" value="/meshdist/cam2/original"/>
        <param name="labelsTopic" value="/meshdist/cam2/projected_pc"/>
        <param name="distanceTopic" value="/meshdist/cam2/projected_pc_distance"/>
    </node>

    <!-- ONLINE LEARNING NODE -->
    <!--node pkg="background_foreground_segmentation"
          type="online_learning.py"
          name="online_learning"
          output="screen">
        <param name="model_gdrive_id" value="$(arg segmentation_model)" />
        <param name="input_height" value="$(arg input_height)" />
        <param name="input_width" value="$(arg input_width)" />
        <rosparam>
        learning_rate: 0.00001
        image_topics:
          - - /pickelhaube_resized/cam0/image
            - /meshdist/cam0/projected_pc
            - /meshdist/cam0/projected_pc_distance
          - - /pickelhaube_resized/cam1/image
            - /meshdist/cam1/projected_pc
            - /meshdist/cam1/projected_pc_distance
          - - /pickelhaube_resized/cam2/image
            - /meshdist/cam2/projected_pc
            - /meshdist/cam2/projected_pc_distance
        segmentation_output_topics:
          - /pickelhaube_resized/cam0/segmented
          - /pickelhaube_resized/cam1/segmented
          - /pickelhaube_resized/cam2/segmented
        pointcloud_topic: /rslidar_points
        label_frequency: 3
        label_options:
          useSuperpixel: true
          foregroundTrustRegion: false
          fgBgThreshold: 50
          numberOfSuperPixel: 400
        </rosparam>
    </node-->

   <!--node pkg="background_foreground_segmentation"
          type="inference_node.py"
          name="inference_node"
          output="screen">
        <param name="model_gdrive_id" value="$(arg segmentation_model)" />
        <param name="input_height" value="$(arg input_height)" />
        <param name="input_width" value="$(arg input_width)" />
        <rosparam>
          image_topics:
            - /pickelhaube_resized/cam0/image
            - /pickelhaube_resized/cam1/image
            - /pickelhaube_resized/cam2/image
          segmentation_output_topics:
            - /pickelhaube_resized/cam0/segmented
            - /pickelhaube_resized/cam1/segmented
            - /pickelhaube_resized/cam2/segmented
          pointcloud_topic: /rslidar_points
        </rosparam>
    </node-->

    <!--node pkg="background_foreground_segmentation"
          type="cloud_filtering_node"
          name="cloud_filter"
          output="screen">
        <param name="cloud_topic" value="/rslidar_points" />
        <param name="cloud_frame" value="rslidar" />
        <param name="filtered_pc_topic" value="/points_filtered" />
        <param name="density_threshold" value="$(arg threshold)" />
        <rosparam param="image_topics">
            - /pickelhaube_resized/cam0/segmented
            - /pickelhaube_resized/cam1/segmented
            - /pickelhaube_resized/cam2/segmented
        </rosparam>
        <rosparam param="caminfo_topics">
            - /pickelhaube_resized/cam0/camera_info
            - /pickelhaube_resized/cam1/camera_info
            - /pickelhaube_resized/cam2/camera_info
        </rosparam>
    </node-->


    <node type="rviz" name="rviz" pkg="rviz" args="-d $(find background_foreground_segmentation)/config/segmented2_icp.rviz" />

    <!-- should be started last for handover to ICP-->
    <include file="$(find background_foreground_segmentation)/launch/pickelhaube/pickelhaube_versavis_from_bag.launch">
        <arg name="rosbag" value="$(arg rosbag)" />
        <arg name="rate" value="0.04" />
    </include>

    <!-- for Realsense: Fix camera_pose_frame to robot-->
    <node pkg="tf2_ros" type="static_transform_publisher" name="tracking_cam_to_rslidar"
				args="0.008 -0.309 0.12 0 0 0 1  /camera_pose_frame /rslidar"/>

    <node name="mapper"
          type="cpt_selective_icp_dyn_node"
          pkg="cpt_selective_icp_dyn"
          output="screen" >
          <!--launch-prefix="tmux split-window" -->
    <rosparam command="load" file="$(find background_foreground_segmentation)/launch/rss_experiments/mapper_parameters_dyn.yaml" />
                <param name="scanTopic" value="/rslidar_points" />
                <param name="odomTopic" value="/camera/odom/sample" />
                <param name="skipScans" value="1" /> 
                <param name="ekfEnable" value="false" />
        <param name="cadTopic" value="mesh_publisher/mesh_out" />
        <param name="icpConfig"
               value="$(find background_foreground_segmentation)/launch/rss_experiments/full_icp.yaml" />
        <param name="selectiveIcpConfig"
               value="$(find background_foreground_segmentation)/launch/rss_experiments/selective_icp.yaml" />
        <param name="inputFiltersConfig"
               value="$(find background_foreground_segmentation)/launch/rss_experiments/input_filters.yaml" />
        <param name="mapPostFiltersConfig"
               value="$(find background_foreground_segmentation)/launch/rss_experiments/map_post_filter.yaml" />
        <param name="mapPreFiltersConfig"
               value="$(find background_foreground_segmentation)/launch/rss_experiments/map_pre_filter.yaml" />
        <param name="path"
                 value="$(find cpt_selective_icp_dyn)" />
    </node>

    <node name="mesh_publisher"
          pkg="cpt_utils"
          type="mesh_publisher_node"
          output="screen"
          launch-prefix="bash -c 'sleep 5; $0 $@'">
        <param name="publish_on_start" value="true"/>
        <param name="default_filename" value="$(arg cad_file)"/>
        <param name="frame_name" value="marker"/>
      </node>
</launch>
